### ISSUE TITLE
Sorting search results

### ISSUE THREAD

**Issue by corbanbrook** [atom/fuzzy-finder#21]
Multiple schemes can be employed to achieve results which are most relevant to the user. Predicting which file a user wants out of a long list of possible matches and presenting it first can help speed up development time/maintain flow and train of thought. Here are some ideas to discuss:
- [ ] Sort least fuzzy results to the top. Fuzziness determined by search term run length in the result. ie. Search term of 'src' would score a higher fuzziness score on a file like .jshintrc than a file called app_src.js or files in the src/ directory. run length of first file is 2 while the second file has a run length of 3.
- [ ] Filename match or directory match. A match on a filename should be sorted at a higher priority than a match on a full path, but what about matches which higher run length on a full path vs less run length on a filename. ie. search of 'src ap' currently displays images/destroy_discard.png higher than src/app.coffee. Although destroy_discard technically has more matched character within the filename than src/app.coffee src/app has more run length. (something to think about)
- [ ] Dot files. Hidden files are less important than non hidden of equal score.
- [ ] Ignored files (.gitignore) can be sorted at lower priority than files with equal score. ie. I want the index.html in my templates/ dir. not my build/ dir because this is in .gitignore. Editing temporary file by mistake can add confusion and lost work.
- [ ] Sort files with most recent last modified date at a higher priority than those of equal score.
- [ ] Filter out/sort to bottom files that are already open.

**Comment by nj** [comment]
:+1: as the current solution is rather useless - and can even be faster to find the file manually

**Comment by matugm** [comment]
:+1:  Would be great if we could get some progress on this.

**Comment by walles** [comment]
Improved sorting / scoring:
https://github.com/atom/fuzzaldrin/pull/22

The above pull request addresses at least some of the issues raised here.

**Comment by mkorfmann** [comment]
Since I'm working on a lot of Rails projects with ActiveAdmin, I'm often annoyed when I end up in an ActiveAdmin file for a particular resource instead of a model file.

I was thinking about improving this by sorting the fuzzy-finder results by usage. I.e. if a files in some folder are worked on more often, they are ranked higher.

I'm happy to implement this experimentally and make a pull request if other people approve of this idea also.

**Comment by Soleone** [comment]
:+1: for some improvements that make finding commonly used files easier. sublime seemed to have done a better job putting the file i actually want to open at the top (using rails here as well)

**Comment by dmnd** [comment]
Just in case further examples are helpful:

![image](https://cloud.githubusercontent.com/assets/4427/9743999/2b4e4856-561e-11e5-8f97-01187fbd7235.png)

**Comment by kevinsimper** [comment]
@jeancroy Did #22 solve this issue?

**Comment by jeancroy** [comment]
There's now an "use Alternate Scoring" option in fuzzy finder that use it.
It address many issue about the search by file name / path.

But it does not cover any knowledge about the file themselves, such as preference for recent / frequent / certain files.

**Comment by r-owen** [comment]
I just tried the Atom Beta with "Use Alternate Scoring" enabled and it's a huge improvement, though still not as good as Sublime Text. I have a project with a huge number of files, including Doxygen generated html files that I rarely want to look at. I tried to find a file named "matchOptimisticB.h". In SublimeText I can type "mob.h" and get the right file as the first suggestion. In Atom Beta it is the ninth choice, preceded by eight html files I have no interest in.

One thing that might help Atom: if the user provides a file type suffix, prefer names that match that suffix exactly over names that use that suffix as a prefix.

Another thing that might help (though I really hope it won't come to this, and it's not needed by Sublime) is to allow the user to disable directory patterns. In my case I might eliminate searches of Doxygen-generated html files and would definitely elimiate .os files (why in the world is it showing binary libraries?).

**Comment by jeancroy** [comment]
Ok please open an issue on fuzzaldrin-plus I can give it a look. I'd need
the result that come before to understand why they are preferred. Also full
path is useful, if private, a mock-up with same length and directory depth.

**Comment by r-owen** [comment]
I just submitted this issue. I hope it helps.

https://github.com/jeancroy/fuzzaldrin-plus/issues/12

Thank you very much for trying to improve Atom’s fuzzy search.

— Russell

**Comment by tnrich** [comment]
Does anyone know if there is an equivalent issue open discussing the Cmd-Shift-P search algorithm?

**Comment by jeancroy** [comment]
you're speaking of command palette? Should already be integrated. If you
have a problem you can try openings a issue on fuzzaldrin-plus repo

**Comment by adamreisnz** [comment]
This is a pretty ancient issue, so I have little hope of improvement arriving any time soon, but here's my two cents:

![image](https://cloud.githubusercontent.com/assets/490562/26134321/e1633240-3b01-11e7-93cf-715e2da67e78.png)

Two things wrong with the way the fuzzy finder currently works both illustrated with the above example.

1) Sort order (as already pointed out in this issue): based on my input, I would really expect the file  **member**/**edit**/**edit**.**html** to show up on top. It does for some reason when I remove the `l` from `html`:

![image](https://cloud.githubusercontent.com/assets/490562/26134356/1ddc312c-3b02-11e7-8416-9e46bb66d4b8.png)

But with the full `html` it suddenly drops to second place which gets rather infuriating after having opened the wrong file several times.

So scoring should somehow take into account how close the search terms are to each other in the filename, and prioritize `edit.html` over `edit-payment.html` unless I include `payment` in my search query.

2) It should really prioritize full word matches rather than scattered letters. If you look at the above example, it actually matches `member` because it's in the app/co**m**pon**e**nts/ad**m**in/mem**ber** folder, instead of simply matching to the `member` part of the path, because that's a whole matching word.

These two tweaks would make the search algorithm a lot stronger.

**Comment by jeancroy** [comment]
hi a@damreisnz , as a curiosity, is this happening with alternate scoring turned on ? There was a strong preference for word "togetherness" in that version.

From screenshot I'm guessing it's not, but if it is I'll add a few of those to test benchmark.
Previous algorithm would take first occurrence of `m` then first `e` then first `m` instead of waiting and trying for `member`

**Comment by adamreisnz** [comment]
@jeancroy yes, it's enabled:

![image](https://cloud.githubusercontent.com/assets/490562/26136645/6034de66-3b11-11e7-8302-2ef7efa63ab4.png)

**Comment by jeancroy** [comment]
Another possibility is that alternate score is used for ranking while classic is used for highlighting.
The whole component below fuzzy finder has been rewritten recently.

----

One feature of the new one is a bias toward file name (vs whole path) when we match file extension exactly I think you are batling against that when you are using keyword from the path but end with extensions

**Comment by adamreisnz** [comment]
Well, my use case as you might deduce from my example is that in a large project, there will be many components. Each component might have an `edit` sub component as in the example, and each of those components will have `edit.html` template, and `edit.js` module, and perhaps `edit.ctrl.js` controller.

So the way I tend to quickly open the file I want, is by specifying the parent component `member`, then the sub component `edit` and then extension if I know there's going to be more than one file.

This usually works fine, but in the above case it was messing it up due to the existence of another similar file. 

I think my use case is fairly common, so I wouldn't expect to be "battling" against the fuzzy finder's system with it.

**Comment by jeancroy** [comment]
You're right on all account, in this case it seems the algorithm just like the `m` of payment. 
I guess the `m` of html manage to count twice, I'll see how to fix that.

Good news is that the issue is more constrained than say lack of prioritizing "full word matches". (Here` -` count as a word boundary)

 
![untitled](https://cloud.githubusercontent.com/assets/12763666/26139983/49208bd6-3aa2-11e7-98f9-d8db304e3134.png)

**Comment by adamreisnz** [comment]
Yeah that looks better in your screenshot, highlighting member properly. And interesting that it likes the `m` in payment and `paykent` is put at the bottom properly. Looks like it's just a few tweaks needed to fix those issues then 👍

**Comment by adamreisnz** [comment]
Looks like in the latest version (just built Atom from master yesterday) there's still some scoring issues. For example this result:

![image](https://user-images.githubusercontent.com/490562/26997451-e751200c-4dcd-11e7-82fb-aac2c5f0d22d.png)

It should not prioritise `cards/club-details.js` over `cards/details.js` for the same reason as above, where it shouldn't prioritise the `edit-payment.html` file.

I did not type a `c` character and it already matched `card`, so it's a bit baffling why it tries to mark the `c` of `club` and give that result a higher score than the more sensible result below that.

Note that when I type `cards` it does prioritise correctly (but still marks the `c` in the second result):

![image](https://user-images.githubusercontent.com/490562/26997484-3db19c60-4dce-11e7-9355-c359a608b47e.png)

I think once a search term has been used/matched in the path, it should not try to match it again for another part of the path.

**Comment by adamreisnz** [comment]
Another example in Atom 1.20 dev where the closest match is not prioritised properly;

![image](https://user-images.githubusercontent.com/490562/28951479-5d87e1c6-791f-11e7-8dfb-0352abc68c89.png)

**Comment by adamreisnz** [comment]
Guys, any activity on this issue please? It's infuriating to keep opening the wrong files because the fuzzy finder sorting logic is off.

VSCode manages to do it correctly, why not Atom? Perhaps it would be worthwhile looking at their algorithm.

![image](https://user-images.githubusercontent.com/490562/31855380-86b1e6bc-b706-11e7-825e-38097545c21b.png)

![image](https://user-images.githubusercontent.com/490562/31855384-95ee7302-b706-11e7-96f4-a4958a936ca7.png)

**Comment by 50Wliu** [comment]
@adamreisnz looks like this was fixed a month ago by @jeancroy but we're running an outdated version of fuzzaldrin-plus.  Will create a PR.
